src/shared/cloudformation/templateSymbolResolver.ts

```diff
@@ -27,45 +22,106 @@ export interface TemplateSymbolProvider {
+        for (let funSymbol of funSymbols) {
+            const found = this.findDescendants([funSymbol], 'Events', vscode.SymbolKind.Module).filter(r =>
+                this.isKind('api', r)
+            )
+            apiSymbols.push(
+                ...found.map(o => ({
+                    // We want the resource name of the Function node (not Api,
+                    // that node is always named "Events".)
+                    name: funSymbol.name,
+                    range: o.range,
+                    kind: kind,
+                }))
+            )
+        }
+        return apiSymbols
     }
 
-    private isFunctionResource(symbol: vscode.DocumentSymbol) {
-        const typeSymbol = _(symbol.children).find({ name: 'Type', kind: vscode.SymbolKind.String })
+    /**
+     * Searches the tree and returns all matching nodes.
+     */
+    private findDescendants(
+        symbols: vscode.DocumentSymbol[],
+        name: string,
+        kind: vscode.SymbolKind
+    ): vscode.DocumentSymbol[] {
+        const found = symbols.filter(v => v.name === name && v.kind === kind)
+        for (let s of symbols) {
+            found.push(...this.findDescendants(s.children, name, kind))
+        }
+        return found
+    }
 
+    private isKind(kind: 'function' | 'api', symbol: vscode.DocumentSymbol) {
+        // Example:
+        // ------------------------------------
+        // Resources:
+        //   ImplicitApiFunction:
+        //     Type: AWS::Serverless::Function
+        //     Properties:
+        //       ...
+        //       Events:             \
+        //         GetHtml:           \
+        //           Type: Api         \ presence of this means
+        //           Properties:       / API instead of FUNCTION
+        //             Path: /        /
+        //             Method: get   /
+        const typeSymbol = this.findDescendants(symbol.children, 'Type', vscode.SymbolKind.String)[0]
+        const wantType =
+            kind === 'function' ? CloudFormation.SERVERLESS_FUNCTION_TYPE : CloudFormation.SERVERLESS_API_TYPE
         if (!typeSymbol) {
             return false
         }
-
         const parsedSymbol = safeLoad(this.symbolProvider.getText(typeSymbol, this.document)) as { Type: string }
-
-        return parsedSymbol.Type === CloudFormation.SERVERLESS_FUNCTION_TYPE
+        return parsedSymbol.Type === wantType
     }
 }
 
-class DefaultSymbolProvider implements TemplateSymbolProvider {
-    public async getSymbols(document: vscode.TextDocument): Promise<vscode.DocumentSymbol[]> {
-        const symbols = await vscode.commands.executeCommand<vscode.DocumentSymbol[]>(
-            'vscode.executeDocumentSymbolProvider',
-            document.uri
-        )
+export class TemplateSymbolProvider {
+    public async getSymbols(document: vscode.TextDocument, waitForSymbols: boolean): Promise<vscode.DocumentSymbol[]> {
+        for (let i = 0; i < 30; i++) {
```

Comment:
where does the magic `30` comes from? do we have any constant timeouts? also do we have a defined structure to do timeouts giving it a function?

Reply:
Do you mean you want to use `timeoutUtils.ts:Timeout` here? I am not sure that API adds clarity here.

Reply:
I guess what I'm fishing for is do we have an equivalent of `waitUntil` and I guess the answer is no

Reply:
PR updated with `waitUntil()` function. Didn't find anything like it in lodash or elswhere in the project.